src/controllers/krate/versions.rs

```diff
@@ -302,4 +355,208 @@ struct PaginatedVersionsAndPublishers {
 struct ResponseMeta {
     total: i64,
     next_page: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    release_tracks: Option<ReleaseTracks>,
+}
+
+#[derive(Debug, Eq, PartialEq, Serialize)]
+struct ReleaseTracks(IndexMap<ReleaseSlug, HighestSemver>);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+enum ReleaseSlug {
+    Minor(u64),
+    Major(u64),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
+struct HighestSemver {
+    highest: semver::Version,
+}
+
+impl ReleaseTracks {
+    // Return the release tracks based on a sorted semver versions iterator (in descending order).
+    // **Remember to** filter out yanked versions manually before calling this function.
+    pub fn from_sorted_semver_iter<'a, I>(versions: I) -> Self
+    where
+        I: Iterator<Item = &'a semver::Version>,
+    {
+        let mut map = IndexMap::new();
+        for num in versions.filter(|num| num.pre.is_empty()) {
+            let key = ReleaseSlug::from_semver(num);
+            let prev = map.last();
+            if prev.filter(|&(k, _)| *k == key).is_none() {
+                map.insert(
+                    key,
+                    HighestSemver {
+                        highest: num.clone(),
+                    },
+                );
+            }
+        }
+
+        Self(map)
+    }
+}
```

Comment:
small nitpick: can we keep the `impl` blocks next to the struct/enum which they are associated with? IMHO it makes it a bit easier to understand the code that way since you don't have to search for impl blocks far away from the struct/enum.

Reply:
Sure!